### PR TITLE
Allow building with cmake (to ease building on windows), version 2

### PR DESCRIPTION
**Pull request by vrabaud**
This is a small rewrite of #16 , mostly to simplify the windows support and modify less code.

### CODE REVIEW

**Comment by vrabaud** [comment]
@trondn, I can take over this pull request until it's merged but I would need you to agree to the Contributor License Agreeement. Is that ok with you ?

**Comment by vrabaud** [comment]
ok, @trondn, the error was on my side, sorry. As asked by the bot, please confirm that you are ok with me committing what you authored, and we are good to go ! Thx !

**Comment by trondn** [comment]
Go for it :)

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    SET(HAVE_STDDEF_H 0)
+ENDIF (NOT HAVE_STDDEF_H)
+
+CHECK_LIBRARY_EXISTS(z zlibVersion "" HAVE_LIBZ)
+CHECK_LIBRARY_EXISTS(lzo2 lzo1x_1_15_compress "" HAVE_LIBLZO2)
+CHECK_LIBRARY_EXISTS(lzf lzf_compress "" HAVE_LIBLZF)
+CHECK_LIBRARY_EXISTS(fastlz fastlz_compress "" HAVE_LIBFASTLZ)
+CHECK_LIBRARY_EXISTS(quicklz qlz_compress "" HAVE_LIBQUICKLZ)
+
+CHECK_CXX_SOURCE_COMPILES("int main(void) { return __builtin_expect(0, 1) ? 1 : 0; }"
```
Comment (alkis) [inline CMakeLists.txt]: Wouldn't the following be enough:
```
CHECK_CXX_SOURCE_COMPILES("int main(void) { return __builtin_expect(0, 1); }" HAVE_BUILTIN_EXPECT)
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        HAVE_BUILTIN_EXPECT)
+
+CHECK_CXX_SOURCE_COMPILES("int main(void) { return __builtin_ctzll(0) ? 1 : 0; }"
```
Comment (alkis) [inline CMakeLists.txt]: Same.

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,151 @@
+PROJECT(Snappy C CXX)
+CMAKE_MINIMUM_REQUIRED(VERSION 3.0)
+
+SET(SNAPPY_MAJOR 1)
+SET(SNAPPY_MINOR 1)
+SET(SNAPPY_PATCHLEVEL 3)
```
Comment (alkis) [inline CMakeLists.txt]: This should be PATCHLEVEL 4.

**Comment by Optiligence** [comment]
you can specify the version in the [`project`](https://cmake.org/cmake/help/latest/command/project.html?highlight=VERSION) command instead of manually setting it
`SOVERSION` will then be set automatically and `WRITE_BASIC_PACKAGE_VERSION_FILE` will also use it automatically.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+SET_TARGET_PROPERTIES(snappy PROPERTIES SOVERSION
+        ${SNAPPY_MAJOR}.${SNAPPY_MINOR}.${SNAPPY_PATCHLEVEL})
+
+SET(INCLUDE_INSTALL_DIR include)
+SET(LIBRARY_INSTALL_DIR lib)
+SET(BINARY_INSTALL_DIR bin)
+
+CONFIGURE_PACKAGE_CONFIG_FILE(cmake/SnappyConfig.cmake.in
+        ${Snappy_BINARY_DIR}/SnappyConfig.cmake
+        INSTALL_DESTINATION ${CMAKE_INSTALL_PREFIX}/lib/Snappy/cmake
```
Comment (Optiligence) [inline CMakeLists.txt]: should be `INSTALL_DESTINATION lib/cmake/Snappy`

**Comment by Optiligence** [comment]
everything `${CMAKE_INSTALL_PREFIX}/lib/cmake` should be `lib/cmake/Snappy`, it currently installs to `<build_dir>/<install_prefix>/<install_prefix>/lib/Snappy/cmake` if a relative `CMAKE_INSTALL_PREFIX` is used

**Comment by Optiligence** [comment]
also, the export set is currently kinda useless because it doesn’t export any targets (and this is basically what https://github.com/google/snappy/pull/16#issuecomment-175704053 was intended to be about)

i’d also vote for using lower case for the CMake code

I added proposed changes here https://github.com/Optiligence/snappy/tree/cmake
I’m not sure how to proceed with this though.

**Comment by vrabaud** [comment]
@Optiligence , I believe I took your comments into account in a new commit (except fro capitalization for now). I also added some older fixes that I had squashed in the older CMakeLists.txt commit (shouldn't have sorry).

**Comment by Optiligence** [comment]
Several parts are different (missing, wrong), you can squash the commits from here https://github.com/Optiligence/snappy/tree/cmake_upper_case instead
although i’d really do the capitalization now too as a baseline instead of sometime later. 
I also added a line on my previous HEAD commit and added another commit regarding static and shared flavors.

**Comment by Optiligence** [comment]
I was also thinking that the output name should contain the ABI version, as it currently does with autotools.

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+CHECK_INCLUDE_FILES("inttypes.h" HAVE_INTTYPES_H)
+CHECK_INCLUDE_FILES("memory.h" HAVE_MEMORY_H)
+CHECK_INCLUDE_FILES("stddef.h" HAVE_STDDEF_H)
+CHECK_INCLUDE_FILES("stdint.h" HAVE_STDINT_H)
+CHECK_INCLUDE_FILES("stdlib.h" HAVE_STDLIB_H)
+CHECK_INCLUDE_FILES("strings.h" HAVE_STRINGS_H)
+CHECK_INCLUDE_FILES("string.h" HAVE_STRING_H)
+CHECK_INCLUDE_FILES("sys/byteswap.h" HAVE_SYS_BYTESWAP_H)
+CHECK_INCLUDE_FILES("sys/endian.h" HAVE_SYS_ENDIAN_H)
+CHECK_INCLUDE_FILES("stdint.h" HAVE_STDINT_H)
```
Comment (KindDragon) [inline CMakeLists.txt]: Duplicate from line 24

**Comment by wesm** [comment]
Any updates here / anything others can do to help? We would love to put up a Snappy windows build on https://github.com/conda-forge/snappy-feedstock

**Comment by alkis** [comment]
Yes we are making progress. We should have arelease with cmake soon, plus
continuous integration.

On Apr 18, 2017 21:23, "Wes McKinney" <notifications@github.com> wrote:

Any updates here / anything others can do to help? We would love to put up
a Snappy windows build on https://github.com/conda-forge/snappy-feedstock

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/google/snappy/pull/29#issuecomment-294952868>, or mute
the thread

**Comment by KindDragon** [comment]
Please add CMake option to build tests https://gist.github.com/KindDragon/b8752d22541a6d9d488e066b6a93a853#file-snappycmakelists-txt-L113

**Comment by alkis** [comment]
I am closing this PR since we have submitted CMake internally and it will be part of the next release.

**Pull request closed without merging**